### PR TITLE
[stable/mattemost-team-edition] Add initial charts for Mattermost Team Edition

### PR DESCRIPTION
**Pull request by cpanato**
**What this PR does / why we need it**:
Add Mattermost Team Edition to Helm charts

**Special notes for your reviewer**:
Please let me know if need any clarifications

### CODE REVIEW

File: incubator/mattermost-team-edition/templates/config.tpl [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        "GatewayAdminSecret": "",
+        "StunURI": "",
+        "TurnURI": "",
+        "TurnUsername": "",
+        "TurnSharedKey": ""
+    },
+    "DataRetentionSettings": {
+        "Enable": false
+    }
+}
```
Comment (jwilander) [inline incubator/mattermost-team-edition/templates/config.tpl]: Might be a good idea to remove any of the enterprise related settings so it's a little cleaner as those won't be needed in the chart

File: incubator/mattermost-team-edition/values.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+image:
+  repository: mattermost/mattermost-team-edition
+  tag: 4.10.1
+
+# Mattermost configuration:
+config:
+  SiteUrl: "http://mattermost.example.com"
+  SiteName: "MMTE"
```
Comment (jwilander) [inline incubator/mattermost-team-edition/values.yaml]: Maybe default this to just `Mattermost`

File: incubator/mattermost-team-edition/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+## TL;DR;
+
+```bash
+$ helm install incubator/mattermost-team-edition
+```
+
+## Introduction
+
+This chart creates a [Mattermost](https://mattermost.com/) deployment on a [Kubernetes](http://kubernetes.io)
```
Comment (jwilander) [inline incubator/mattermost-team-edition/README.md]: Might be a good idea to stress that it's team edition so:

>This chart creates a Mattermost deployment...

to

>This chart creates a Mattermost Team Edition deployment...

**Review by jwilander — COMMENTED** [review]
Awesome! Just a few minor comments

File: incubator/mattermost-team-edition/README.md (hunk 3 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/README.md]: done

File: incubator/mattermost-team-edition/templates/config.tpl (hunk 1 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/config.tpl]: done

File: incubator/mattermost-team-edition/values.yaml (hunk 2 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/values.yaml]: done

**Review by jwilander — APPROVED** [review]
LGTM

**Comment by cpanato** [comment]
hello
@lachie83 @linki I know you guys are busy, but when we expect some reviews for this pr? 😃

**Comment by clkao** [comment]
Thanks for the chart! I wonder if it is possible to define a configmap generated by the `config` key in values, which would override arbitrary keys in config.json (maybe with jq's object merge).  With that we don't need to change the chart for new configurable options, for example gitlab SSO integration.

File: incubator/mattermost-team-edition/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            httpGet:
+              path: /api/v4/system/ping
+              port: {{ .Values.service.internalPort }}
+          volumeMounts:
+          - mountPath: /mattermost/config/config.json
+            name: config-json
+            subPath: config.json
+          resources:
+{{ toYaml .Values.resources | indent 12 }}
+      volumes:
```
Comment (clkao) [inline incubator/mattermost-team-edition/templates/deployment.yaml]: should there be an option for /mattermost/data be persistent with a pvc?

Comment (cpanato) [reply incubator/mattermost-team-edition/templates/deployment.yaml]: ~~for the data we are using minio~~

Comment (cpanato) [reply incubator/mattermost-team-edition/templates/deployment.yaml]: ahh, sorry minio is for the other edition. Will add the persistence for that 😄 thanks for the catch!

**Review by clkao — COMMENTED** [review]
For people using hosted sql, it'd be useful if this chart provides extraContainers and extraVolumes to configure cloud proxy.  See the [keycloak chart]( https://github.com/kubernetes/charts/tree/master/stable/keycloak#using-google-cloud-sql-proxy) for example

File: incubator/mattermost-team-edition/values.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  ingress: false
+
+# Mattermost configuration:
+config:
+  SiteUrl: "http://mattermost.example.com"
+  SiteName: "Mattermost"
+  FilesAccessKey:
+  FilesSecretKey:
+  FileBucketName:
+  SMTPHost:
```
Comment (clkao) [inline incubator/mattermost-team-edition/values.yaml]: it seem this should be `SMTPServer` according to config.tpl

**Review by clkao — CHANGES_REQUESTED** [review]

**Comment by cpanato** [comment]
hello @clkao thanks for all review and feedback!
this part is not clear to me what need to be done, can you help me on that?
>  I wonder if it is possible to define a configmap generated by the config key in values, which would override arbitrary keys in config.json (maybe with jq's object merge). With that we don't need to change the chart for new configurable options, for example gitlab SSO integration.

File: incubator/mattermost-team-edition/values.yaml (hunk 5 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/values.yaml]: done

File: incubator/mattermost-team-edition/templates/deployment.yaml (hunk 4 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/deployment.yaml]: done

**Comment by cpanato** [comment]
@clkao for this request:
> For people using hosted sql, it'd be useful if this chart provides extraContainers and extraVolumes to configure cloud proxy.

I think it is better to add in a follow up, we need to change how we handle our configuration and this change will allow better manipulation. For now I will put this as TODO. is that ok for you?

**Comment by clkao** [comment]
I've added a few things including postgresql support, recreate on config changes, and the required hooks for cloud sql proxy.  Let me push somewhere for you to see if the changes can be incorporated.

**Comment by cpanato** [comment]
@clkao Hello, I was reading more about the cloud sql proxy that you mentioned, and the problem for that it is limited only to Google cloud i thought that was more generic.

**Comment by clkao** [comment]
Hi @cpanato I made a few improvements in separate commits so it should be easier for you to cherry-pick: https://github.com/kubernetes/charts/compare/pull/5987/head...clkao:mattermost

**Comment by cpanato** [comment]
hi @clkao sorry for the delay I was in vacations. I cherry pick all suggestions thanks!!

File: incubator/mattermost-team-edition/templates/service.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  selector:
+    app: {{ template "mattermost-team-edition.name" . }}
+  type: {{ .Values.service.type }}
+  ports:
+  - port: {{ .Values.service.externalPort }}
+    targetPort: {{ .Values.service.internalPort }}
+    protocol: TCP
+    name: {{  template "mattermost-team-edition.name" . }}
+  selector:
+    app: {{ template "mattermost-team-edition.name" . }}
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/service.yaml]: Add `release` label to selector.

File: incubator/mattermost-team-edition/templates/tests/mattermost-config-test.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{template  "mattermost-team-edition.fullname" .}}-tests
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/tests/mattermost-config-test.yaml]: Add standard labels.

File: incubator/mattermost-team-edition/templates/pvc.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.persistence.data.enabled }}
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/pvc.yaml]: Add standard labels.

File: incubator/mattermost-team-edition/templates/configmap-config.yaml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{template "mattermost-team-edition.fullname" .}}-config-json
+  labels:
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+    app: {{ template "mattermost-team-edition.fullname" . }}
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/configmap-config.yaml]: Add `release` and `heritage` labels.

File: incubator/mattermost-team-edition/templates/service.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "mattermost-team-edition.name" . }}
+  labels:
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/service.yaml]: Add `release` and `heritage` labels.

File: incubator/mattermost-team-edition/templates/service.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    app: {{ template "mattermost-team-edition.name" . }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+spec:
+  selector:
+    app: {{ template "mattermost-team-edition.name" . }}
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/service.yaml]: Add `release` label to selector.

File: incubator/mattermost-team-edition/templates/tests/mattermost-test.yaml [hunk 12]
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: "{{template  "mattermost-team-edition.fullname" .}}-test-{{ randAlphaNum 5 | lower }}"
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/tests/mattermost-test.yaml]: Add standard labels.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please use a consistent list indentation style throughout the chart. Either use zero or two spaces.

Why does the chart have dependencies on kube-lego and nginx-ingress? You don't install these on a per-chart basis. You usually have multiple services use the same ingress controller. Also, the ingress that's included is tied to nginx-ingress. What if someone wants to use a different ingress controller? kube-lego is deprecated. I would rather use cert-manager, but again, not on a per-chart basis. I'd suggest you remove these dependencies.

File: incubator/mattermost-team-edition/templates/service.yaml (hunk 6 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/service.yaml]: done

File: incubator/mattermost-team-edition/templates/tests/mattermost-config-test.yaml (hunk 7 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/tests/mattermost-config-test.yaml]: done

File: incubator/mattermost-team-edition/templates/pvc.yaml (hunk 8 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/pvc.yaml]: done

File: incubator/mattermost-team-edition/templates/configmap-config.yaml (hunk 9 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/configmap-config.yaml]: done

File: incubator/mattermost-team-edition/templates/service.yaml (hunk 10 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/service.yaml]: done

File: incubator/mattermost-team-edition/templates/service.yaml (hunk 11 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/service.yaml]: done

File: incubator/mattermost-team-edition/templates/tests/mattermost-test.yaml (hunk 12 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/tests/mattermost-test.yaml]: done

**Comment by cpanato** [comment]
@unguiculus thanks so much for your review and the ingress/kubelego makes sense. removed.

ptal

File: incubator/mattermost-team-edition/Chart.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
+description: Mattermost Team Edition server.
+name: mattermost-team-edition
+version: 0.1.0
+appVersion: 5.0.0
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/Chart.yaml]: The image used is 5.1.1

File: incubator/mattermost-team-edition/values.yaml [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    ##
+    # storageClass:
+    accessMode: ReadWriteOnce
+  ## This volume is used to store git data and other project files.
+  ## ref: https://docs.gitlab.com/omnibus/settings/configuration.html#storing-git-data-in-an-alternative-directory
+  ##
+
+# Mattermost configuration:
+config:
+  SiteUrl: "http://mattermost.example.com"
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/values.yaml]: Convention is to use lower camel case.

File: incubator/mattermost-team-edition/templates/deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+{{- end }}
+          resources:
+{{ toYaml .Values.resources | indent 12 }}
+{{- with .Values.extraContainers }}
+{{ tpl . $ | indent 8 }}
+{{- end }}
+      volumes:
+        - name: config-json
+          configMap:
+            name: {{template "mattermost-team-edition.fullname" .}}-config-json
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/deployment.yaml]: Please be consistent. Add a space after `{{` am  before `}}`. Apply throughout the chart.

File: incubator/mattermost-team-edition/templates/ingress.yaml [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "mattermost-team-edition.fullname" . }}
+  labels:
+    chart: {{ template "mattermost-team-edition.chart" . }}
+    app: {{ template "mattermost-team-edition.fullname" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  annotations:
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/ingress.yaml]: Don't set nginx-ingress specific annotations by default. Instead provide an option to configure annotations. You may add these defaults to `values.yaml` and comment them out.

File: incubator/mattermost-team-edition/templates/service.yaml [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+metadata:
+  name: {{ template "mattermost-team-edition.name" . }}
+  labels:
+    chart: {{ template "mattermost-team-edition.chart" . }}
+    app: {{ template "mattermost-team-edition.fullname" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  selector:
+    app: {{ template "mattermost-team-edition.name" . }}
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/templates/service.yaml]: Add release label to selector.

File: incubator/mattermost-team-edition/templates/service.yaml (hunk 6 above)
Comment (unguiculus) [reply incubator/mattermost-team-edition/templates/service.yaml]: I don't see it?

**Review by unguiculus — CHANGES_REQUESTED** [review]
Again, please fix list indentation as mentioned above.

File: incubator/mattermost-team-edition/Chart.yaml (hunk 13 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/Chart.yaml]: done

File: incubator/mattermost-team-edition/values.yaml (hunk 14 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/values.yaml]: done

File: incubator/mattermost-team-edition/templates/deployment.yaml (hunk 15 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/deployment.yaml]: done

File: incubator/mattermost-team-edition/templates/service.yaml (hunk 17 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/service.yaml]: done

File: incubator/mattermost-team-edition/templates/service.yaml (hunk 6 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/service.yaml]: my fault added only in the label.

File: incubator/mattermost-team-edition/templates/ingress.yaml (hunk 16 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/templates/ingress.yaml]: done, thanks!

File: incubator/mattermost-team-edition/values.yaml [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+service:
+  type: ClusterIP
+  externalPort: 8065
+  internalPort: 8065
+
+ingress:
+  enabled: false
+  path: /
+  annotations:
+    kubernetes.io/ingress.class: nginx
```
Comment (unguiculus) [inline incubator/mattermost-team-edition/values.yaml]: Should be commented out

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/mattermost-team-edition/values.yaml (hunk 18 above)
Comment (cpanato) [reply incubator/mattermost-team-edition/values.yaml]: done

File: stable/mattermost-team-edition/README.md [hunk 19]
```diff
@@ -0,0 +1,110 @@
+# Mattermost Team Edition
+
+[Mattermost](https://mattermost.com/) is a hybrid cloud enterprise messaging workspace that brings your messaging and tools together to get more done, faster.
+
+## TL;DR;
+
+```bash
+$ helm install incubator/mattermost-team-edition
```
Comment (unguiculus) [inline stable/mattermost-team-edition/README.md]: Fix references to incubator.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/mattermost-team-edition/README.md (hunk 19 above)
Comment (cpanato) [reply stable/mattermost-team-edition/README.md]: done @unguiculus

File: stable/mattermost-team-edition/templates/deployment.yaml [hunk 20]
```diff
@@ -0,0 +1,76 @@
+apiVersion: apps/v1beta2
```
Comment (unguiculus) [inline stable/mattermost-team-edition/templates/deployment.yaml]: Can you use `apps/v1`?

File: stable/mattermost-team-edition/values.yaml [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+  # externalDriverType: "postgres" #or mysql
+  # externalConnectionString: "postgres://<USERNAME>:<PASSWORD>@<HOST>:5432/<DATABASE_NAME>?sslmode=disable&connect_timeout=10"
+  # for mysql: "<USERNAME>:<PASSWORD>@tcp(<HOST>:3306)/<DATABASE_NAME>?charset=utf8mb4,utf8&readTimeout=30s&writeTimeout=30s"
+
+mysql:
+  enabled: true
+  mysqlRootPassword: root_password
+  mysqlUser: mmuser
+  mysqlPassword: mmpasswd
+  mysqlDatabase: mattermost
```
Comment (unguiculus) [inline stable/mattermost-team-edition/values.yaml]: Don't set a default password.

File: stable/mattermost-team-edition/values.yaml [hunk 22]
```diff
@@ -0,0 +1,113 @@
+# Default values for mattermost-team-edition.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+image:
+  repository: mattermost/mattermost-team-edition
+  tag: 5.2.1
+  imagePullPolicy: Always
```
Comment (unguiculus) [inline stable/mattermost-team-edition/values.yaml]: IfNotPresent

File: stable/mattermost-team-edition/values.yaml [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+  ## This volume persists generated data from users, like images, attachments...
+  ##
+  data:
+    enabled: true
+    size: 10Gi
+    ## If defined, volume.beta.kubernetes.io/storage-class: <storageClass>
+    ## Default: volume.alpha.kubernetes.io/storage-class: default
+    ##
+    # storageClass:
+    accessMode: ReadWriteOnce
```
Comment (unguiculus) [inline stable/mattermost-team-edition/values.yaml:23]: What if I run multiple replicas? How would Mattermost behave?

**Review by unguiculus — CHANGES_REQUESTED** [review]
Again, please fix list indentation.

File: stable/mattermost-team-edition/values.yaml (hunk 23 above)
Comment (cpanato) [reply stable/mattermost-team-edition/values.yaml:23]: For team edition, you can run just one instance.

File: stable/mattermost-team-edition/templates/deployment.yaml (hunk 20 above)
Comment (cpanato) [reply stable/mattermost-team-edition/templates/deployment.yaml]: done

File: stable/mattermost-team-edition/values.yaml (hunk 21 above)
Comment (cpanato) [reply stable/mattermost-team-edition/values.yaml]: done

File: stable/mattermost-team-edition/values.yaml (hunk 22 above)
Comment (cpanato) [reply stable/mattermost-team-edition/values.yaml]: done

File: stable/mattermost-team-edition/values.yaml (hunk 23 above)
Comment (cpanato) [reply stable/mattermost-team-edition/values.yaml:23]: made note in the Readme.md as limitation.

File: stable/mattermost-team-edition/templates/pvc.yaml [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+  labels:
+    chart: {{ template "mattermost-team-edition.chart" . }}
+    app: {{ template "mattermost-team-edition.fullname" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  annotations:
+  {{ range $key, $value := .Values.persistence.data.annotations }}
+    {{ $key }}: {{ $value | quote }}
+  {{ end }}
+spec:
```
Comment (unguiculus) [inline stable/mattermost-team-edition/templates/pvc.yaml:15]: Add `selector.matchLabels` including `app` and `release` labels.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/mattermost-team-edition/templates/pvc.yaml (hunk 24 above)
Comment (cpanato) [reply stable/mattermost-team-edition/templates/pvc.yaml:15]: @unguiculus If enable the matchLabels for the PVC the PVC is in pending forever. I guess then we need to create the PV and add the labels there to work

screen shot 2018-09-17 at 09 35 45

Comment (unguiculus) [reply stable/mattermost-team-edition/templates/pvc.yaml:15]: I was wrong. Dynamic provision is not possible if `matchLabels` is set.

**Review by unguiculus — APPROVED** [review]

**Comment by cpanato** [comment]
@unguiculus thanks for all the review!

**Comment by consideRatio** [comment]
@cpanato wieeeeeeeeeeeeeeee! :D I'm excited about this!

@clkao you appear all over =D

**Pull request merged**